File: typescript/exercises.ts
import { open } from "node:fs/promises"

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let counts: Map<bigint, bigint> = new Map()
  let remaining = amount
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination)
    remaining %= denomination
  }
  return counts
}

export function firstThenApply<T, U>(items: T[], predicate: (item: T) => boolean, consumer: (item: T) => U): U | undefined {
  var first_element: T | undefined = items.find(predicate)

  return first_element !== undefined ? consumer(first_element): undefined
}


export function* powersGenerator(base: bigint): Generator<bigint>{
  for (let exponentiation = 1n; ; exponentiation *= base)
    yield exponentiation
}

export async function meaningfulLineCount(filename: string){
  let count: number = 0
  const file = await open(filename, "r")
  for await (const line of file.readLines()){
    const trimmed: string = line.trim()
    if (trimmed && !trimmed.startsWith("#")) {
      count++
    }
  }
  return count
}

interface Sphere {
  kind: "Sphere"
  radius: number
}

interface Box {
  kind: "Box"
  width: number
  length: number
  depth: number
}

export type Shape = Sphere | Box

export function surfaceArea(shape: Shape): number {

  switch(shape.kind){
    case "Sphere":
      return 4 * Math.PI * shape.radius ** 2
    case "Box":
      return 2 * shape.width * shape.length + 2 * shape.length * shape.depth + 2 * shape.width * shape.depth
  }

}

export function volume(shape: Shape): number {
  switch(shape.kind){
    case "Sphere":
      return 4/3 * Math.PI * shape.radius ** 3
    case "Box":
      return shape.width * shape.length * shape.depth
  }
}


export interface BinarySearchTree<T>{
  size(): number
  insert(value: T): BinarySearchTree<T>
  contains(value: T): boolean
  inorder(): Iterable<T>
}

export class Empty<T> implements BinarySearchTree<T> {
  size(): number {
    return 0
  }

  insert(value: T): BinarySearchTree<T> {
    return new Node(this, value, this);
  }

  contains(value: T): boolean {
      return false
  }

  *inorder(): Iterable<T> {
  
  }

  toString(): String {
    return "()"
  }
}

export class Node<T> implements BinarySearchTree<T> {
  left: BinarySearchTree<T> | Empty<T>
  right: BinarySearchTree<T> | Empty<T>
  value: T

  constructor(left: BinarySearchTree<T> | Empty<T>, value: T, right: BinarySearchTree<T> | Empty<T>){
    this.left = left;
    this.right = right;
    this.value = value;
  }

  size(): number {
    return this.left.size() + 1 + this.right.size()
  }

  insert(value: T): BinarySearchTree<T> {
    if (value < this.value){
      return new Node(this.left.insert(value), this.value, this.right)
    }
    else if (value > this.value){
      return new Node(this.left, this.value, this.right.insert(value))
    }
    else {
      return new Node(this.left, this.value, this.right)
    }
  }

  contains(value: T): boolean {
      return this.value === value || this.right.contains(value) || this.left.contains(value)
  }

  *inorder(): Iterable<T> {
    yield* this.left.inorder()
    yield this.value
    yield* this.right.inorder()
  }

  toString(): String {
    return (("(" + this.left.toString() + String(this.value) + this.right.toString() + ")").replace(/\(\)/g, ""))
  }
}

File: ocaml/exercises.ml
exception Negative_Amount

let change amount =
  if amount < 0 then
    raise Negative_Amount
  else
    let denominations = [25; 10; 5; 1] in
    let rec aux remaining denominations =
      match denominations with
      | [] -> []
      | d :: ds -> (remaining / d) :: aux (remaining mod d) ds
    in
    aux amount denominations

let first_then_apply (array: 'a list) (predicate: 'a -> bool) (consumer: 'a -> 'b): 'b =
  match List.find_opt predicate array with
  | Some x -> consumer x
  | None -> None

let powers_generator base = 
  let rec generate_from power () =
    Seq.Cons (power, generate_from (power * base))
  in
  generate_from 1

let meaningful_line_count filename =
  let meaningful_line line = 
    let trimmed = String.trim line in
    String.length trimmed > 0 && not (String.starts_with ~prefix:"#" trimmed)
  in

  let the_file = open_in filename in
  let finally () = close_in the_file in

  let rec count_lines count =
    try
      let line = input_line the_file in 
      if meaningful_line line then
        count_lines (count + 1)
      else
        count_lines count
    with
    | End_of_file -> count
  in

  Fun.protect ~finally (fun () -> count_lines 0)

type shape = 
  | Sphere of float
  | Box of float * float * float

let volume s = 
  match s with
  | Sphere r -> Float.pi *. (r ** 3.0) *. 4.0 /. 3.0
  | Box (l, w, h) -> l *. w *. h

let surface_area s =
  match s with
  | Sphere r -> 4.0 *. (r**2.0) *. Float.pi
  | Box (l, w, h) -> 2.0 *. l *. w +. 2.0 *. l *. h +. 2.0 *. w *. h

type 'a binary_search_tree = 
  | Empty
  | Node of 'a binary_search_tree * 'a * 'a binary_search_tree

let rec size tree =
  match tree with
  | Empty -> 0
  | Node (left, _, right) -> 1 + size left + size right

let rec contains data tree =
  match tree with
  | Empty -> false
  | Node (left, value, right) -> data == value || contains data left || contains data right

let rec insert comparable tree =
  match tree with
  | Empty -> Node (Empty, comparable, Empty)
  | Node (left, value, right) ->
    if comparable < value then 
      Node (insert comparable left, value, right)
    else if comparable > value then 
      Node (left, value, insert comparable right)
    else
      Node (left, value, right)

let rec inorder tree =
    match tree with
    | Empty -> []
    | Node (left, value, right) -> inorder left @ [value] @ inorder right